### PR TITLE
Use async/await instead of async package

### PR DESCRIPTION
**Pull request by rekmarks**
_Note:_ We should test this out with the extension and internal packages before merging and publishing.

Replaces the `async` NPM package with use of Promises and native `async`/`await` functionality.
Should, and must be, functionally equivalent.

Removing `async` will decrease this package's bundle size by ~73%: https://bundlephobia.com/result?p=json-rpc-engine@5.1.8

### CODE REVIEW

File: src/asMiddleware.js [hunk 1]
```diff
@@ -1,19 +1,30 @@
 'use strict'
 
 module.exports = function asMiddleware (engine) {
-  return function engineAsMiddleware (req, res, next, end) {
```
Comment (Gudahtt) [inline src/asMiddleware.js:4]: This does change the return type of this function, from `undefined` to `Promise`. Is this change required to remove `async`? I don't immediately see the connection.

File: src/index.js [hunk 2]
```diff
@@ -70,115 +69,113 @@ module.exports = class RpcEngine extends SafeEventEmitter {
...
-    // down stack of middleware, call and collect optional allReturnHandlers
-    async.mapSeries(this._middleware, eachMiddleware, completeRequest)
+    // go down stack of middleware, call and collect optional allReturnHandlers
+    for (const middleware of this._middleware) {
+      if (!isComplete) {
+        await runMiddleware(middleware)
+      }
+    }
+
+    const returnHandlers = allReturnHandlers.reverse()
```
Comment (Gudahtt) [inline src/index.js]: Previously these were never returned if an error occurred, but now they are returned.

Comment (Gudahtt) [reply src/index.js]: Maybe this is fine, since they still aren't actually used in the error case 🤔

Comment (Gudahtt) [reply src/index.js]: Oh, wait, maybe they are. `res.result` isn't being deleted in the error scenario anymore, so `_checkForCompletion` would no longer throw.

**Review by Gudahtt — CHANGES_REQUESTED** [review]

File: src/asMiddleware.js (hunk 1 above)
Comment (rekmarks) [reply src/asMiddleware.js:4]: The reason for the change is in order to `await engine._runMiddlewareDown` on the subsequent line, but I believe it is convertible to `.then`/`.catch`. I'll give that a shot.

Comment (rekmarks) [reply src/asMiddleware.js:4]: f919617

File: src/index.js (hunk 2 above)
Comment (rekmarks) [reply src/index.js]: Hm, per the spec, the return handlers should called in case of `res.error`, as long as `_checkForCompletion` does _not_ throw, which it will only do if `(!res.result && !res.error) || isComplete`. See: https://github.com/MetaMask/json-rpc-engine/blob/remove-async/test/basic.spec.js/#L265

I did add back the `res.result` deletion in the middleware in f919617, but that would never cause `_checkForCompletion` to throw as long as `res.error` was set.

**Comment by rekmarks** [comment]
@Gudahtt summary of changes in f919617:

- Make middleware returned by `asMiddleware` synchronous, by using `.then`/`.catch` instead of `async`/`await`
- Restore deletion of `res.result` in middleware `end` function
- Ensure errors from `_checkForCompletion`, return handlers, or some unexpected source are set on the response
  - I believe the failure to do this was a bug. See `_handle`
- Cleanup
  - Rename a couple of functions in `index.js` to more specific names

File: src/index.js (hunk 2 above)
Comment (Gudahtt) [reply src/index.js]: Oh I see what you mean. Yeah that makes sense.

**Review by Gudahtt — APPROVED** [review]
LGTM!
